Msvc2017

ericmehl opened the pull request:
Here's the Cortex PR for Windows compatibility. Mostly unchanged from what @boberfly did earlier. I rebased it on the latest and sorted a couple final details.

It's successfully building on MSVC 2017 but I'm sorting out building on Linux due to unrelated issues stemming more from the dependencies and my Ubuntu build environment. So I _think_ it should be fine for Linux generally but haven't been able to verify that for sure.

boberfly commented:
Hey Eric looks like you're having fun carrying the torch here for the Windows/MSVC2017 stuff. I'll review a few things once I work out how to review properly on a per-line/file/commit basis. Cheers.

boberfly commented on python/IECore/ClassLoader.py:
```diff
@@ -215,7 +215,7 @@ def defaultLoader( cls, envVar ) :
 				msg( Msg.Level.Warning, "ClassLoader.defaultLoader", "Environment variable %s not set." % envVar )
 
-			loader = cls( SearchPath( os.path.expandvars( sp ) ) )
+			loader = cls( SearchPath( os.path.expandvars( sp ), os.pathsep ) )
```
What I recommend here is to remove the separator entirely, because in a previous PR John and I discussed this and decided to do this internally in SearchPath without needing the separator on the end. Here is what I mean: https://github.com/ImageEngine/cortex/blob/master/include/IECore/SearchPath.h#L58

ericmehl replied:
Ahh, I think I get it now. I had seen that thread but took it to conclude with using separate declarations for the few files affected. I'll adjust as you outline. The idea is to push the path separators back to the Python UI code instead, correct?

boberfly replied:
@ericmehl I think just doing ```SearchPath( os.path.expandvars( sp ) )``` should suffice, the C++ code internally will use the right separator internally, so the Python UI code shouldn't need to worry about any separator passed as an argument.

boberfly commented:
I see that you're using the bat file in there, @donboie recently made it so that the CMakeLists.txt 'should' work from the contrib/cmake directory now (at least on Linux) so that the bat file doesn't need to first copy it out of there to the root source dir. Have you been able to make it work like that? I haven't tried myself under MSVC, but we could remove the xcopy then and that message at the top.

boberfly commented on src/IECoreScene/TransformStack.cpp:
```diff
@@ -32,6 +32,8 @@
 //
 //////////////////////////////////////////////////////////////////////////
 
+#include<algorithm>
```
Probably needs a space here

ericmehl commented:
I'll make some tweaks later today and try and track down what's causing the build to fail on Travis as well.

ericmehl commented on contrib/IECoreAppleseed/include/IECoreAppleseed/ProgressTileCallback.h:
```diff
@@ -56,7 +56,7 @@ class IECOREAPPLESEED_API ProgressTileCallback : public renderer::TileCallbackBa
 
 		void release() override;
 
-		void on_tile_end( const renderer::Frame *frame, const size_t tileX, const size_t tileY ) override;
```
I'm not familiar with clang so not sure on that one. I'll dig up the exact issue MSVC was failing on - from memory I think it was saying something along the lines that it wasn't actually overriding anything but I'll have a closer look.

ericmehl commented on contrib/cmake/CMakeLists.txt:
```diff
@@ -303,6 +303,7 @@ if( ${WITH_IECORE_VDB} )
     file( GLOB IECOREVDB_PY_FILES ${CORTEX_SOURCE_DIR}/python/IECoreVDB/*.py )
     add_library( IECoreVDB SHARED ${IECOREVDB_CXX_FILES} ${IECOREVDB_H_FILES} )
     target_link_libraries( IECoreVDB IECore IECoreScene ${BASE_LIBS} ${OpenVDB_LIBRARIES} )
+    target_compile_options( IECoreVDB PRIVATE -DOPENVDB_3_ABI_COMPATIBLE )
```
I could make it optional or maybe the root cause is the way VDB is being built in the dependencies? I'll have a look at that and maybe we can remove all the -DOPENVDB_3_ABI_COMPATIBLE.

ericmehl commented:
I'll go back and pull some of the exact errors that were driving the changes to have a more precise conversation.

I do have a successful TeamCity setup here and I hope it wouldn't be too hard to make a similar setup on Appveyor, at least to get builds going. I can lend a hand with the Appveyor effort if you like.

I don't know about running the tests so might need to make that a second step effort.

dboogert commented:
> I'll go back and pull some of the exact errors that were driving the changes to have a more precise conversation.

I know this is a pain but some additional information on the exact problem encountered would give context for the change. 

> I do have a successful TeamCity setup here and I hope it wouldn't be too hard to make a similar setup on Appveyor, at least to get builds going. I can lend a hand with the Appveyor effort if you like.

This would be great! I really don't know how much time we can devote (if any) but would be great if the reference Windows build was completely available in the public with CI so we'd be much less likely to break the build with our Linux development. Are you able to create a PR with your build scripts in contrib/TeamCity so I can take a look?

> I don't know about running the tests so might need to make that a second step effort.

Of course. I've got it on my todo list to get the cmake build execute tests on Linux so I hope this might help on Windows.

ericmehl commented:
> I'll go back and pull some of the exact errors that were driving the changes to have a more precise conversation.
> 
> I know this is a pain but some additional information on the exact problem encountered would give context for the change.
> 
I totally understand, and you might well catch some things or have better methods than what I sorted out.

>     I do have a successful TeamCity setup here and I hope it wouldn't be too hard to make a similar setup on Appveyor, at least to get builds going. I can lend a hand with the Appveyor effort if you like.
> 
> This would be great! I really don't know how much time we can devote (if any) but would be great if the reference Windows build was completely available in the public with CI so we'd be much less likely to break the build with our Linux development. Are you able to create a PR with your build scripts in contrib/TeamCity so I can take a look?

I'll add the Kotlin scripts that TeamCity exported, hopefully tonight. It looks like that's the most useful as everything is very web-UI configuration driven in TC. It's not much more than triggering the build command really, but there are a few environment variables to set. 

> 
>     I don't know about running the tests so might need to make that a second step effort.
> 
> Of course. I've got it on my todo list to get the cmake build execute tests on Linux so I hope this might help on Windows.

ericmehl commented:
I just added a somewhat rough TeamCity build script. Not sure how familiar you are with TC but it's mostly configured through the web based interface with settings being spread out in number of parts of the GUI. I use it essentially as an automated way of plugging in commands to the terminal, nothing too fancy.

The Kotlin scripts it let me download look encompassing - I removed a reference to my own server and a user / password combo for my GitHub login.

There are a couple triggers left over that refer to my Gaffer Dependencies build and the Gaffer build that are used or kicked off respectively.

To summarize the function of it:
1. set the environment variables in https://github.com/hypothetical-inc/cortex/blob/98b8e5ebfd93b9332d2954f13111249c1966b839/contrib/TeamCity/Cortex/buildTypes/Cortex_WindowsCortexBuild.kt#L17-L24
2. run contrib/cmake/buildCortexWindows.bat
3. take everything from the temp directory (a directory TC supplies as a variable and is guaranteed to be clean when the build starts) and zip it up as the delivered result
4. If any of the build commands return non-zero it fails the build and supplies the log via the web interface

I've been working on additional dependencies cleanup so haven't gotten into Cortex yet. It's next on my list, and when I get the hang of that I'll clean up the Gaffer build that I currently have working but not pretty enough to do a PR from yet.

ericmehl commented:
I just about have a rebased and organized update ready to submit for consideration. I'm hung up on one last problem, specifically with https://github.com/hypothetical-inc/cortex/blob/554bbe6097b1d392e0cdd164a24084cffa08f2d6/src/IECore/IndexedIOAlgo.cpp#L84

MSVC is giving me this compiler error: 

> Severity	Code	Description	Project	File	Line	Suppression State
Error	C2664	'void IECore::IndexedIO::read(const IECore::IndexedIO::EntryID &,unsigned short &) const': cannot convert argument 2 from 'long *' to 'float *&'	IECore	m:\cortex\src\iecore\indexedioalgo.cpp	84	

My understanding of templates is a work in process, so please bear with me. I can't track down what is referencing that Reader class with a "long" type.  I solved the problem in MSVC (hypothetical-inc@2ac33a2b607bf4151836991c0c35f26bc6d9b9c) by adding an overload to handle long, but that caused an error when building on gcc.

Maybe the error gcc gives will be of help? When building with the hypothetical-inc@2ac33a2b607bf4151836991c0c35f26bc6d9b9c7 commit:

> scons: *** [src/IECore/CompoundObject.os] Error 1
[16:26:39]W:	 [Step 2/2] In file included from include/IECore/Object.h:39:0,
[16:26:39]W:	 [Step 2/2]                  from include/IECore/CompoundObject.h:39,
[16:26:39]W:	 [Step 2/2]                  from include/IECore/CompoundParameter.h:38,
[16:26:39]W:	 [Step 2/2]                  from src/IECore/CompoundParameter.cpp:35:
[16:26:39]W:	 [Step 2/2] include/IECore/IndexedIO.h:252:16: error: 'virtual void IECore::IndexedIO::write(const EntryID&, const int64_t*, long unsigned int)' cannot be overloaded
[16:26:39]W:	 [Step 2/2]    virtual void write(const IndexedIO::EntryID &name, const int64_t *x, unsigned long arrayLength) = 0;
[16:26:39]W:	 [Step 2/2]                 ^~~~~
[16:26:39]W:	 [Step 2/2] include/IECore/IndexedIO.h:222:16: error: with 'virtual void IECore::IndexedIO::write(const EntryID&, const long int*, long unsigned int)'
[16:26:39]W:	 [Step 2/2]    virtual void write(const IndexedIO::EntryID &name, const long *x, unsigned long arrayLength) = 0;
[16:26:39]W:	 [Step 2/2]                 ^~~~~
[16:26:39]W:	 [Step 2/2] include/IECore/IndexedIO.h:330:16: error: 'virtual void IECore::IndexedIO::write(const EntryID&, const int64_t&)' cannot be overloaded
[16:26:39]W:	 [Step 2/2]    virtual void write(const IndexedIO::EntryID &name, const int64_t &x) = 0;
[16:26:39]W:	 [Step 2/2]                 ^~~~~
[16:26:39]W:	 [Step 2/2] include/IECore/IndexedIO.h:305:16: error: with 'virtual void IECore::IndexedIO::write(const EntryID&, const long int&)'
[16:26:39]W:	 [Step 2/2]    virtual void write(const IndexedIO::EntryID &name, const long &x) = 0;
[16:26:39]W:	 [Step 2/2]                 ^~~~~
[16:26:39]W:	 [Step 2/2] include/IECore/IndexedIO.h:401:16: error: 'virtual void IECore::IndexedIO::read(const EntryID&, int64_t*&, long unsigned int) const' cannot be overloaded
[16:26:39]W:	 [Step 2/2]    virtual void read(const IndexedIO::EntryID &name, int64_t *&x, unsigned long arrayLength) const  = 0;
[16:26:39]W:	 [Step 2/2]                 ^~~~
[16:26:39]W:	 [Step 2/2] include/IECore/IndexedIO.h:371:16: error: with 'virtual void IECore::IndexedIO::read(const EntryID&, long int*&, long unsigned int) const'
[16:26:39]W:	 [Step 2/2]    virtual void read(const IndexedIO::EntryID &name, long *&x, unsigned long arrayLength) const = 0;
[16:26:39]W:	 [Step 2/2]                 ^~~~
[16:26:39]W:	 [Step 2/2] include/IECore/IndexedIO.h:479:16: error: 'virtual void IECore::IndexedIO::read(const EntryID&, int64_t&) const' cannot be overloaded
[16:26:39]W:	 [Step 2/2]    virtual void read(const IndexedIO::EntryID &name, int64_t &x) const  = 0;
[16:26:39]W:	 [Step 2/2]                 ^~~~
[16:26:39]W:	 [Step 2/2] include/IECore/IndexedIO.h:454:16: error: with 'virtual void IECore::IndexedIO::read(const EntryID&, long int&) const'
[16:26:39]W:	 [Step 2/2]    virtual void read(const IndexedIO::EntryID &name, long &x) const = 0;

Any tips on tracking down that incorrect template implementation?

danieldresser-ie commented:
Hmm, I'm not sure what the solution is, but it looks like the reason why GCC doesn't like the override for long is because there is already an override for int64_t, and GCC seems to be considering long and int64 to be synonymous.

MSVC on the other hand seems to find the case where a "long" is passed in, and think that the closest match is "float".  Is there no more context from the MSVC error about what code is triggering this template expansion?

danieldresser-ie commented:
Oh, this looks like there is a pretty simple explanation.

We've specialized for int64_t, but we use the specialization for long.  This works on GCC because it considers these types synonyms, but MSVC doesn't.

Can you try replacing those lines with:
```
		case IndexedIO::Long:
			Handler<int64_t, Callback>().handleValue( src, dst, entry, c );
			break;
		case IndexedIO::LongArray:
			Handler<int64_t, Callback>().handleArray( src, dst, entry, c );
			break;
```

This seems more correct anyway - IndexIO::Long is actually about how the data is saved to file, so if for some reason we were loading IndexIO on a platform where "long" != "int64_t", we would still always want to deserialize IndexedIO::Long as int64_t.

ericmehl commented:
Wonderful! That did the trick. Thanks a lot for helping sort that out, it makes sense to me now.

johnhaddon commented:
I ran up against a similar problem with long and IndexedIOAlgo when compiling for OSX. We merged this fix to master already, and I imagine it should also fix things for you :

https://github.com/ImageEngine/cortex/commit/ccb8746f6cc13352507129f30143afb80d3fa73b

Note that my solution was simply to remove the support for Long entirely rather than switch to `int64_t`. I had to do some software archeology to figure this out because the Long and LongArray cases have been obsolete for a very long time; they date back to when we were using 32 bit platforms, and in moving to 64 bit platforms we switched to doing things properly with int64 etc. A weird detail appears to be that on the 32 bit platforms where we did use Long, a long was actually only 32 bits anyway, so we had literally no need for it...

ericmehl commented:
Ahh, that does make it easier. I rebased so we should be up to date now. I want to check that Gaffer itself it building on both Windows and Linux and I'll merge in my changes for consideration on this PR.
